Feature: Preprocess JVM properties at app start to adjust to system environment.

infeo opened the pull request:
Fixes #2838. Relates to https://github.com/cryptomator/cryptomator/pull/2885#issuecomment-1587093110

This PR introduces a preprocessing of JVM properties at application start to react system environments settings. The preprocessing is done in way which will be compatible with the JDK 21 preview feature of String templates.

To substitute a string in a property, it needs to be enclosed in ~~`@( )`~~ `@{}` and may only contain the characters A-Z, a-z, 0-9 and _.  Only the keywords `appdir`,`appdata`, `localappdata` and `userhome` are supported for now.

TODO:
* [x] adjust buildscripts
* [x] replace `~` by `@(userhome)` in code

infeo commented:
After merge, issue https://github.com/cryptomator/cryptomator/issues/2853 can also be fixed.

overheadhunter commented on src/main/java/org/cryptomator/common/PropertiesPreprocessor.java:
```diff
@@ -0,0 +1,66 @@
+public class PropertiesPreprocessor {
+
+	private static final Logger LOG = LoggerFactory.getLogger(PropertiesPreprocessor.class);
+	private static final Pattern TEMPLATE = Pattern.compile("@\\((.+)\\)");
```
This Pattern is greedy, it would not only match `@(hello)` but also `@(hello) world)`. To be 100% exact, we would have to add an escape mechanism (`@(hello\) world)`), which would probably not work with regex but requires a manually written state machine.

To keep it simple, we might want to use a regex which disallows certain characters (e.g. `@\\(([^)]+)\\)`), or even simpler just allows certain chars (`@\\((\w)\\)`).

---

On a separate note: Maybe we should not use `@()` but rather `@{}`.  `{}` don't need to be quoted, making the regex more legible, furthermore these characters are less common in the actual property (maybe, we want to eventually support `some.property.with.@{thing.getName()}`)

infeo replied:
Only allow certain characters is a good idea!

Regarding the token: Actually, `{` and `}` must be quoted as well because they are used in character classes and quantifiers. But the latter part about calling functions is a good point!

overheadhunter commented on src/main/java/org/cryptomator/common/PropertiesPreprocessor.java:
```diff
@@ -0,0 +1,66 @@
+					case "appdata" -> ENV.get("APPDATA");
+					case "userhome" -> System.getProperty("user.home");
+					default -> {
+						LOG.warn("Found unknown keyword {} in property value {}. Keyword is not replaced.", match.group(), value);
```
I'd include the `@()` to make the log more user-friendly

```suggestion
						LOG.warn("Found unknown variable @({}) in property value {}. Keyword is not replaced.", match.group(), value);
```

infeo replied:
Changed in https://github.com/cryptomator/cryptomator/pull/2957/commits/4065e15aa1f0b099a900b056ae26c5b3d143d03b

overheadhunter commented on src/main/java/org/cryptomator/common/PropertiesPreprocessor.java:
```diff
@@ -0,0 +1,66 @@
+	private static final Pattern TEMPLATE = Pattern.compile("@\\((.+)\\)");
+	private static final LoggingEnvironment ENV = new LoggingEnvironment(System.getenv(), LOG);
+
+	public static void run() {
+		var properties = System.getProperties();
+		properties.stringPropertyNames().stream() //
+				.filter(s -> s.startsWith("cryptomator.")) //
+				.forEach(key -> {
+					var value = properties.getProperty(key);
+					var newValue = process(value);
+					if(! value.equals(newValue)) {
+						LOG.info("Changed property {} from {} to {}.", key, value, newValue);
+					}
+					properties.setProperty(key, newValue);
+				});
+		LOG.info("Preprocessed cryptomator properties.");
+	}
```
Please don't actually "change" the properties. Rather run the `process` step when reading from properties inside `Environment.java`

infeo replied:
As already said over a different channel: This won't be sufficient for underlying libraries, since they cannot access the Environment class of Cryptomator. An idea would be to add another API to integrations-api to read such properties.

My argument is, that we are only changing properties of the Cryptomator namespace, we do it at the main entry point before loading any additional classes and the Properties API even allows such changes. Additionally, the properties are fixed at application start. If we evaluate them dynamically, we have less guarantees.

infeo replied:
For the sake of agility, we will use the current seek and replace approach. I'll open a seperate issue to put this coupling between Cryptomator and its libraries into a more formal (api) way.

overheadhunter commented on src/main/java/org/cryptomator/launcher/Cryptomator.java:
```diff
@@ -63,7 +64,7 @@ public static void main(String[] args) {
 			return;
 		}
-
+		PropertiesPreprocessor.run();
```
see previous comment.

overheadhunter commented on src/main/java/org/cryptomator/common/PropertiesPreprocessor.java:
```diff
@@ -0,0 +1,66 @@
+public class PropertiesPreprocessor {
+
+	private static final Logger LOG = LoggerFactory.getLogger(PropertiesPreprocessor.class);
+	private static final Pattern TEMPLATE = Pattern.compile("@\\{(\\w+)}");
```
So we need to quote `{` but not `}`? That looks odd.

infeo replied:
Well, you can try it yourself and see if you get it working 😉

overheadhunter reviewed (changes requested):
Can you add a test for the `process` function with various inputs and their expected outputs as parameters?

infeo commented:
> Can you add a test for the process function with various inputs and their expected outputs as parameters?

Done in https://github.com/cryptomator/cryptomator/pull/2957/commits/2c0474ec46f77649adb01fb93f46160ee5e57546
____

I had to circumvent a problem making the situation less clean: The logDir is specified by a property, which gets resolved at class initiatlization (due to creating a logger for `Cryptomator.java`). I moved the `PropertiesPreprocesser::run` directive now to the constructor of the `Environment` object and initialize it also in `Cryptomator.java` to make it more obvious.

Addiitonally, i changed on Windows some locations from appdata to [localappdata](https://learn.microsoft.com/en-us/windows/apps/design/app-settings/store-and-retrieve-app-data#local-app-data).

overheadhunter commented on src/main/java/org/cryptomator/common/Environment.java:
```diff
@@ -34,7 +34,10 @@ public class Environment {
 	private static final String TRAY_ICON_PROP_NAME = "cryptomator.showTrayIcon";
 
-	private Environment() {}
+	private Environment() {
+		//hack, needed for logging directory to be setup correctly
+		PropertiesPreprocessor.run();
+	}
```
This gets invoked after static initialization, so the Logger already exists at this point.

overheadhunter commented on src/test/java/org/cryptomator/common/PropertiesPreprocessorTest.java:
```diff
@@ -0,0 +1,24 @@
+
+public class PropertiesPreprocessorTest {
+
+	@ParameterizedTest
+	@CsvSource(value = """
+			org.example.@{mytest1}.test, org.example.@{mytest1}.test
+			@{only*word*digits*under_score\\},@{only*words*digits*under_score\\}
+			C:\\Users\\@{appdir}\\dir, C:\\Users\\\\dir
+			@{@{appdir}},@{}
+			Longer @{appdir} text with @{appdir}., Longer  text with .
+			""")
+	public void test(String propertyValue, String expected) {
+		var result = PropertiesPreprocessor.process(propertyValue);
+		Assertions.assertEquals(result, expected);
+	}
```
This test has no control over the value of `appdir`, making its result system-dependent.

Can you instead add an overloaded function for `process` which accepts a second argument (the environment), which can get mocked?

infeo replied:
Fixed in https://github.com/cryptomator/cryptomator/pull/2957/commits/ebc60c4fb37421442e6b23fa3f3f38c8204b26a3

overheadhunter reviewed (changes requested):
I dislike the static `ENV` in the main class. Let's have a call to discuss alternative solutions for the initialization order problem.

infeo commented:
After some suggestions from @overheadhunter, I refactored the code.

`PropertiesPreprocessor` is replaced a decorator of the system Properties. `getProperties()` is overriden to lazily substitute variables in the requested property, if it starts with `cryptomator.`. The main class `Cryptomator.java` has now a static initalizer block, in which the default system properties are replaced by its decorator via `System.setProperties()`.

The log configuration class now directly access the decorated properties instead of `Environment`, because it loaded before the fully initiallization of the main class.

overheadhunter commented on src/main/java/org/cryptomator/common/PropertiesDecorator.java:
```diff
@@ -0,0 +1,171 @@
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+public class PropertiesDecorator extends Properties {
```
either make this class package-private (preferred) or if you think this will ever be used in any other place, make the constructor public as well

overheadhunter commented on src/test/java/org/cryptomator/common/EnvironmentTest.java:
```diff
@@ -22,41 +22,7 @@ public class EnvironmentTest {
 		env = Mockito.spy(Environment.getInstance());
-		Mockito.when(env.getHomeDir()).thenReturn(Path.of("/home/testuser"));
 	}
-
-	@Test
-	@DisplayName("cryptomator.settingsPath=~/.config/Cryptomator/settings.json:~/.Cryptomator/settings.json")
-	public void testSettingsPath() {
-		System.setProperty("cryptomator.settingsPath", "~/.config/Cryptomator/settings.json:~/.Cryptomator/settings.json");
-
-		List<Path> result = env.getSettingsPath().toList();
-		MatcherAssert.assertThat(result, Matchers.hasSize(2));
-		MatcherAssert.assertThat(result, Matchers.contains(Paths.get("/home/testuser/.config/Cryptomator/settings.json"), //
-				Paths.get("/home/testuser/.Cryptomator/settings.json")));
-	}
-
-	@Test
-	@DisplayName("cryptomator.ipcSocketPath=~/.config/Cryptomator/ipc.socket:~/.Cryptomator/ipc.socket")
-	public void testIpcSocketPath() {
-		System.setProperty("cryptomator.ipcSocketPath", "~/.config/Cryptomator/ipc.socket:~/.Cryptomator/ipc.socket");
-
-		List<Path> result = env.ipcSocketPath().toList();
-		MatcherAssert.assertThat(result, Matchers.hasSize(2));
-		MatcherAssert.assertThat(result, Matchers.contains(Paths.get("/home/testuser/.config/Cryptomator/ipc.socket"), //
-				Paths.get("/home/testuser/.Cryptomator/ipc.socket")));
-	}
-
-	@Test
-	@DisplayName("cryptomator.integrationsWin.keychainPaths=~/AppData/Roaming/Cryptomator/keychain.json")
-	public void testKeychainPath() {
-		System.setProperty("cryptomator.integrationsWin.keychainPaths", "~/AppData/Roaming/Cryptomator/keychain.json");
-
-		List<Path> result = env.getKeychainPath().toList();
-		MatcherAssert.assertThat(result, Matchers.hasSize(1));
-		MatcherAssert.assertThat(result, Matchers.contains(Paths.get("/home/testuser/AppData/Roaming/Cryptomator/keychain.json")));
-	}
-
```
when removing all these tests, please make sure to add equivalent tests to `LazyProcessedPropertiesTest.java`

infeo replied:
What are these tests for? My reading is, that we want to ensure that "~" gets resolved correctly for those specific properties and that the property might contain multiple paths.

With this PR, it is not the task of `Environment.java` to resolve anything. Only parsing the content to paths. I'll readd the tests in a "slim" form again.

overheadhunter replied:
What I meant to say is this: Yes, they can be removed from `EnvironmentTest.java`. But instead I would like to see them in `SubstitutingPropertiesTest.java`. Yes, you have already added tests for `process()`, but the ones I would like to keep are more "end to end":

```java
@ParameterizedTest(name = "{0}={1} -> {0}={2}")
@DisplayName("Replace @{userhome} during getProperty()")
@CsvSource(quoteCharacter = '"', textBlock = """
		cryptomator.settingsPath, "@{userhome}/.config/Cryptomator/settings.json:@{userhome}/.Cryptomator/settings.json", "/home/.config/Cryptomator/settings.json:/home/.Cryptomator/settings.json"
		cryptomator.ipcSocketPath, "@{userhome}/.config/Cryptomator/ipc.socket:@{userhome}/.Cryptomator/ipc.socket", "/home/.config/Cryptomator/ipc.socket:/home/.Cryptomator/ipc.socket"
		not.cryptomator.not.substituted, "@{userhome}/foo", "@{userhome}/foo"
		cryptomator.no.placeholder.found, "foo/bar", "foo/bar"
		// ...
		"""})
public void test(String key, String raw, String substituted) {
	var delegate = Mockito.mock(Properties.class);
	Mockito.doReturn("/home").when(delegate).getProperty("user.home");
	Mockito.doReturn(raw).when(delegate).getProperty(key);
	var inTest = new SubstitutingProperties(delegate, Map.of());

	var result = inTest.getProperty(key);

	Assertions.assertEquals(substituted, result);
}

// same for `appdir`, since this is a different "Source"
```

infeo replied:
Implemented in https://github.com/cryptomator/cryptomator/pull/2957/commits/ee82965ef5d7e069ba00f691508e267930850a8d

overheadhunter commented on src/main/java/org/cryptomator/common/LazyProcessedProperties.java:
```diff
@@ -0,0 +1,73 @@
+		this.env = systemEnvironment;
+	}
+
+	@Override
+	public String getProperty(String key) {
+		var value = delegate.getProperty(key);
+		if (key.startsWith("cryptomator.") && value != null) {
+			return process(value);
+		} else {
+			return value;
+		}
+	}
+
+	@Override
+	public String getProperty(String key, String defaultValue) {
+		var value = delegate.getProperty(key, defaultValue);
+		if (key.startsWith("cryptomator.") && value != null) {
+			return process(value);
+		} else {
+			return value;
+		}
+	}
```
please add tests with 100% branch coverage (i.e. both `cryptomator.*` as well as `example.*` keys as well as value present and absent)

overheadhunter replied:
(while trivial enough to see that it will not cause problems, we need to protect against regressions, as this may affect the whole application startup sequence)

infeo replied:
done in https://github.com/cryptomator/cryptomator/pull/2957/commits/cee0486d71037233b30672d0b5341b7e6417b3cb

overheadhunter commented on src/main/java/org/cryptomator/common/LazyProcessedProperties.java:
```diff
@@ -0,0 +1,73 @@
+					case "appdir" -> resolveFrom("APPDIR", Source.ENV);
+					case "appdata" -> resolveFrom("APPDATA", Source.ENV);
+					case "localappdata" -> resolveFrom("LOCALAPPDATA", Source.ENV);
+					case "userhome" -> resolveFrom("user.home", Source.PROPS);
```
if we expose `user.home` via `Environment.java`, we can remove `Source`, making the code simpler.

infeo replied:
I assume some misunderstanding happend here: `Source.ENV` points to the OS environment (retrievable by `System.getenv()`), not the Cryptomator internal class

overheadhunter commented on src/main/java/org/cryptomator/common/LazyProcessedProperties.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.Properties;
+import java.util.regex.Pattern;
+
+public class LazyProcessedProperties extends PropertiesDecorator {
```
Last but not least: I don't like the "Lazy" part in the name. Just because the previous solution eagerly replaced all properties, this class isn't doing anything "lazy" in particular. The class itself behaves "normal".

In other words: Better rename to e.g. `SubstitutingProperties` (or if we make `Environment` our only source, maybe even `EnvSubstProperties`, a little tribute to the awesome Linux `envsubst` command).

infeo replied:
Renamed to `SubstitutingProperties` in https://github.com/cryptomator/cryptomator/pull/2957/commits/79eea4814ec361745fc8dfe3ddfc307e705b38f9

overheadhunter commented on src/main/java/org/cryptomator/common/SubstitutingProperties.java:
```diff
@@ -0,0 +1,69 @@
+		return result != null ? result : defaultValue;
+	}
+
+	//visible for testing
```
we can use `@org.jetbrains.annotations.VisibleForTesting` as well

infeo replied:
Added in https://github.com/cryptomator/cryptomator/pull/2957/commits/29d63a0a83b7399ba65bd07ff57bf28e5a770ffd

overheadhunter commented on src/test/java/org/cryptomator/common/SubstitutingPropertiesTest.java:
```diff
@@ -0,0 +1,151 @@
+		public void test(String propertyValue, String expected) {
+			SubstitutingProperties inTest = new SubstitutingProperties(Mockito.mock(Properties.class), Map.of("APPDIR", "foobar"));
+			var result = inTest.process(propertyValue);
+			Assertions.assertEquals(result, expected);
```
should be:

```suggestion
			Assertions.assertEquals(expected, result);
```

(same in various other places)

infeo replied:
Fixed in  https://github.com/cryptomator/cryptomator/pull/2957/commits/29d63a0a83b7399ba65bd07ff57bf28e5a770ffd
